File: 정수빈/Programmers/Level2/괄호회전하기.py
# https://school.programmers.co.kr/learn/courses/30/lessons/76502
# 설명: 문자열을 회전시키면서 올바른 괄호 문자열인지 확인하는 문제입니다.
# 가장 바깥에 문자열을 회전하는 loop
# stack을 사용하여 올바른 괄호 확인

def solution(s):
    answer = 0
    rotated = s
    stack = []
    for i in range(len(s)):
        rotated = s[i:]+s[:i]
        for c in rotated:
            only_closed = False
            if c == "{" or c == "(" or c == "[":
                stack.append(c)
            elif len(stack) <= 0:
                only_closed = True
                continue
            elif stack[-1] == "{" and c == "}":
                stack.pop(-1)
            elif stack[-1] == "(" and c == ")":
                stack.pop(-1)
            elif stack[-1] == "[" and c == "]":
                stack.pop(-1)
            else:
                continue
        if len(stack) == 0 and not only_closed:
            answer += 1
        only_closed = False
        stack = []
    return answer

File: 곽경진/Programmers/Level2/괄호_회전하기 .java
/*
프로그래머스 괄호 회전 시키기
https://school.programmers.co.kr/learn/courses/30/lessons/76502

해야할 것
1. 올바른 괄호문자열인지 확인하기 -> '}]) 가 먼저 나온다면, 올바른 괄호문자열이 아님'
2. string 값 회전시키기

    1. string 값을 length 만큼 반복시키며 마지막 문자열이 맨 앞으로 오게 만듦
    2. 올바른 괄호 문자열인지 확인
        2-1. 올바른 문자열이면 +1
        2-2. 아니라면 계속진행
*/
import java.util.Stack;

class Solution {
    public int solution(String s) {
        
        int answer = 0;
        
        if (s.length()%2 == 1 || !parenthesesSet(s)){
            return 0;
        }
        
        for (int i = 0; i < s.length(); i++) {
            if (parenthesesCheck(s)) {
                answer++;
            }
            s = stringShift(s);
        }
        
        return answer;
    }
    
    public String stringShift(String str) {
        return str.substring(1) + str.charAt(0);
    }
    
    public boolean parenthesesCheck(String str){
        Stack<Character> stack = new Stack<>();

        for (char ch : str.toCharArray()) {
            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch);
            }
            else if (ch == ')' || ch == '}' || ch == ']') {
                if (stack.isEmpty()) {
                    return false;
                }
                char openBracket = stack.pop();
                if (!isMatchingPair(openBracket, ch)) {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
    
     private static boolean isMatchingPair(char open, char close) {
        return (open == '(' && close == ')') ||
               (open == '{' && close == '}') ||
               (open == '[' && close == ']');
    }
    
    public boolean parenthesesSet(String str) {
    return str.chars().filter(c -> c == '(').count() == str.chars().filter(c -> c == ')').count()
        && str.chars().filter(c -> c == '{').count() == str.chars().filter(c -> c == '}').count()
        && str.chars().filter(c -> c == '[').count() == str.chars().filter(c -> c == ']').count();
    }
}
